### PR TITLE
Add new query builder and shared scan CTE

### PR DESCRIPTION
**Pull request by angelini**
The goal of this PR is twofold:

- Don't include path predicates if they are trivial and empty
- Add a shared CTE which filters out all objects except the ones that may be relevant for this query using only version numbers and the project ID

We'd like to try this to ensure that we begin by doing a simple index scan to get just the possible rows and then building out the following CTEs from that. Right now we scan the objects index twice, once for our updated objects and again for possibly deleted objects.

And removing the path predicates will help make the query both simpler to read and should allow Postgres to make simpler index queries.

----

To make this easier to manage I built a really simple `queryBuilder` type. The main utility is that instead of tracking `$1`, `$2`, ... placeholders during query buildup we can use nice indicators like `__project__` or `__start_version__`. This is necessary because we now sometimes fully omit query args and figuring out the arg index was getting super messy.

----

https://blog.jooq.org/10-cool-sql-optimisations-that-do-not-depend-on-the-cost-model/#top4

It seems Postgres might not be removing trivial predicates as I had expected it to 😞

### CODE REVIEW

File: internal/db/queryBuilder.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+		ignorePatterns := []string{}
+		for _, ignore := range qb.objectQuery.Ignores {
+			ignorePatterns = append(ignorePatterns, fmt.Sprintf("%s%%", ignore))
+		}
+
+		args = append(args, ignorePatterns)
+	}
+
+	for idx, name := range argNames {
+		query = strings.ReplaceAll(query, name, fmt.Sprintf("$%d", idx+1))
```
Comment (airhorns) [inline internal/db/queryBuilder.go:183]: You'd think sql might have named query parameters by now, but I guess not :(

File: internal/db/queryBuilder.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+func (qb *queryBuilder) build() (string, []any) {
+	var query string
+
+	if qb.vrange.From == 0 {
+		query = qb.queryWithoutRemovals()
+	} else {
+		query = qb.queryWithRemovals()
+	}
```
Comment (airhorns) [inline internal/db/queryBuilder.go:156]: 👍 but lets make sure we still review the explain for the one with the removals bit in there cause my hunch is that the join with the subquery is the actual issue

**Review by airhorns — APPROVED** [review]
🤘

Do you have a local explain that looks better to you?

**Comment by angelini** [comment]
Query:

```
EXPLAIN ANALYZE WITH possible_objects AS (
	
		SELECT *
		FROM dl.objects
		WHERE project = 3937
		 AND (
		   -- live objects
		   (
	         start_version > 1
		     AND start_version <= 870
		     AND (stop_version IS NULL OR stop_version > 870)
		   )
		   OR
		   -- removed objects
			(
		     start_version <= 870
		     AND stop_version > 1
		     AND stop_version <= 870
		   )
		 )
	
	), updated_objects AS (
	
		SELECT o.path, o.mode, o.size, c.bytes, o.packed, false AS deleted
		FROM possible_objects o
		LEFT JOIN dl.contents c
		 ON o.hash = c.hash
		 
		WHERE o.project = 3937
			AND o.start_version > 1
			AND o.start_version <= 870
			AND (o.stop_version IS NULL OR o.stop_version > 870)
			
			
		ORDER BY o.path
	
	), removed_objects AS (
	
		SELECT o.path, o.mode, 0 AS size, ''::bytea as bytes, o.packed, true AS deleted
		FROM possible_objects o
		WHERE o.project = 3937
			AND o.start_version <= 870
			AND o.stop_version > 1
			AND o.stop_version <= 870
			
			AND (
				-- Skip removing files if they are in the updated_objects list
				(RIGHT(o.path, 1) != '/' AND o.path NOT IN (SELECT path FROM updated_objects))
				OR
				-- Skip removing empty directories if any updated_objects are within that directory
				(RIGHT(o.path, 1) = '/' AND NOT EXISTS (SELECT true FROM updated_objects WHERE STARTS_WITH(path, o.path)))
			)
		ORDER BY o.path
	
	)

	SELECT path, mode, size, bytes, packed, deleted
	FROM updated_objects
	UNION
	SELECT path, mode, size, bytes, packed, deleted
	FROM removed_objects;
```

Result:
```
Unique  (cost=1758.18..1758.21 rows=2 width=82) (actual time=8.843..9.104 rows=307 loops=1)
  CTE possible_objects
    ->  Bitmap Heap Scan on objects  (cost=116.13..1715.85 rows=354 width=131) (actual time=0.481..1.424 rows=3072 loops=1)
          Recheck Cond: ((((project = 3937) AND (stop_version IS NULL)) OR ((project = 3937) AND (start_version > 1) AND (start_version <= 870) AND (stop_version > 870))) OR ((project = 3937) AND (start_version <= 870) AND (stop_version > 1) AND (stop_version <= 870)))
          Filter: ((start_version <= 870) AND (((start_version > 1) AND ((stop_version IS NULL) OR (stop_version > 870))) OR ((stop_version > 1) AND (stop_version <= 870))))
          Rows Removed by Filter: 288
          Heap Blocks: exact=293
          ->  BitmapOr  (cost=116.13..116.13 rows=422 width=0) (actual time=0.445..0.448 rows=0 loops=1)
                ->  BitmapOr  (cost=59.91..59.91 rows=136 width=0) (actual time=0.227..0.228 rows=0 loops=1)
                      ->  Bitmap Index Scan on identical_objects_idx  (cost=0.00..17.45 rows=90 width=0) (actual time=0.047..0.047 rows=373 loops=1)
                            Index Cond: ((project = 3937) AND ((stop_version IS NULL) = true))
                      ->  Bitmap Index Scan on objects_idx  (cost=0.00..42.28 rows=46 width=0) (actual time=0.179..0.180 rows=189 loops=1)
                            Index Cond: ((project = 3937) AND (start_version > 1) AND (start_version <= 870) AND (stop_version > 870))
                ->  Bitmap Index Scan on objects_idx  (cost=0.00..56.13 rows=286 width=0) (actual time=0.218..0.218 rows=2798 loops=1)
                      Index Cond: ((project = 3937) AND (start_version <= 870) AND (stop_version > 1) AND (stop_version <= 870))
  CTE updated_objects
    ->  Sort  (cost=19.20..19.21 rows=1 width=502) (actual time=5.372..5.384 rows=274 loops=1)
          Sort Key: o_1.path
          Sort Method: quicksort  Memory: 195kB
          ->  Nested Loop Left Join  (cost=0.55..19.19 rows=1 width=502) (actual time=0.843..5.126 rows=274 loops=1)
                ->  CTE Scan on possible_objects o_1  (cost=0.00..10.62 rows=1 width=81) (actual time=0.789..2.362 rows=274 loops=1)
                      Filter: ((start_version > 1) AND (start_version <= 870) AND ((stop_version IS NULL) OR (stop_version > 870)) AND (project = 3937))
                      Rows Removed by Filter: 2798
                ->  Index Scan using contents_pkey on contents c  (cost=0.55..8.57 rows=1 width=505) (actual time=0.010..0.010 rows=1 loops=274)
                      Index Cond: (hash = o_1.hash)
  ->  Sort  (cost=23.12..23.13 rows=2 width=82) (actual time=8.842..8.883 rows=1174 loops=1)
        Sort Key: updated_objects.path, updated_objects.mode, updated_objects.size, updated_objects.bytes, updated_objects.packed, updated_objects.deleted
        Sort Method: quicksort  Memory: 345kB
        ->  Append  (cost=0.00..23.11 rows=2 width=82) (actual time=5.374..8.218 rows=1174 loops=1)
              ->  CTE Scan on updated_objects  (cost=0.00..0.02 rows=1 width=82) (actual time=5.374..5.484 rows=274 loops=1)
              ->  Subquery Scan on "*SELECT* 2"  (cost=23.04..23.07 rows=1 width=82) (actual time=2.512..2.668 rows=900 loops=1)
                    ->  Sort  (cost=23.04..23.05 rows=1 width=78) (actual time=2.510..2.543 rows=900 loops=1)
                          Sort Key: o.path
                          Sort Method: quicksort  Memory: 151kB
                          ->  CTE Scan on possible_objects o  (cost=0.02..23.03 rows=1 width=78) (actual time=0.119..2.261 rows=900 loops=1)
                                Filter: ((start_version <= 870) AND (stop_version > 1) AND (stop_version <= 870) AND (project = 3937) AND ((("right"(path, 1) <> '/'::text) AND (NOT (hashed SubPlan 3))) OR (("right"(path, 1) = '/'::text) AND (NOT (SubPlan 4)))))
                                Rows Removed by Filter: 2172
                                SubPlan 3
                                  ->  CTE Scan on updated_objects updated_objects_1  (cost=0.00..0.02 rows=1 width=32) (actual time=0.001..0.029 rows=274 loops=1)
                                SubPlan 4
                                  ->  CTE Scan on updated_objects updated_objects_2  (cost=0.00..0.02 rows=1 width=0) (actual time=0.053..0.053 rows=1 loops=9)
                                        Filter: starts_with(path, o.path)
                                        Rows Removed by Filter: 217
Planning Time: 1.034 ms
Execution Time: 9.417 ms
```

File: internal/db/queryBuilder.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+			AND o.stop_version <= __stop_version__
+			%s
+			AND (
+				-- Skip removing files if they are in the updated_objects list
+				(RIGHT(o.path, 1) != '/' AND o.path NOT IN (SELECT path FROM updated_objects))
+				OR
+				-- Skip removing empty directories if any updated_objects are within that directory
+				(RIGHT(o.path, 1) = '/' AND NOT EXISTS (SELECT true FROM updated_objects WHERE STARTS_WITH(path, o.path)))
+			)
+		ORDER BY o.path
```
[the remark this answers is not in the document]
Comment (angelini) [reply internal/db/queryBuilder.go:103]: There is some logic further downstream when packing files that needs them to be ordered.

**Pull request merged**